### PR TITLE
refactor: use internal directory for private apis

### PR DESCRIPTION
**Pull request by ryancurrah**
Use internal dir for private APIs while improving the public APIs for consumption. Also adding more tests. https://github.com/jaimeteb/chatto/issues/8

### CODE REVIEW

**Comment by ryancurrah** [comment]
This is an early preview of the refactor.

**Comment by jaimeteb** [comment]
Great I understand the concept of internal packages now, I had only seen it in a few projects before and I thought it was just a name, but it is very useful that Go itself uses the name to "hide" certain packages.

Of course, the `query` and `extension` packages (and `fsm` to some extent) should be public, in order to develop the extensions.

Also, some `bot` functions should be public, for example, `ServeBot`, that way you can build a "custom" bot, for example (this uses an old Chatto version):

```go
package main

import "github.com/jaimeteb/chatto/bot"

func main() {
        path := "data/"
        port := 4770
        bot.ServeBot(&path, &port)
}
```

Directory structure:

```
.
├── app.yaml
├── chatto_gae.go
├── data
│   ├── clf.yml
│   └── fsm.yml
├── go.mod
└── go.sum
```

**Comment by ryancurrah** [comment]
Gotcha. I will fix that so people can run the bot.

File: fsm/fsm.go [hunk 1]
```diff
@@ -57,64 +179,88 @@ type FSM struct {
...
-
-	if fsmDomain.TransitionTable[tupleFromAny] == nil {
-		if fsmDomain.TransitionTable[tupleCmdAny] == nil {
-			transition = fsmDomain.TransitionTable[tupleNormal] // There is no transition "From Any" with cmd, nor "Cmd Any"
-			tuple = tupleNormal
-		} else {
-			transition = fsmDomain.TransitionTable[tupleCmdAny] // There is a transition "Cmd Any"
-			tuple = tupleCmdAny
+// TODO: Document what the ExecuteCmd does
+func (m *FSM) ExecuteCmd(cmd, question string, fsmDomain *Domain) (answers []query.Answer, extension string) {
```
Comment (ryancurrah) [inline fsm/fsm.go:183]: I broke the `ExecuteCmd` down into smaller pieces. It was large and did quite a few things. I think this way it makes it easier to understand.

File: fsm/fsm.go [hunk 2]
```diff
@@ -57,64 +179,88 @@ type FSM struct {
...
-			if r, err := regexp.Compile(slot.Regex); err == nil {
-				match := r.FindAllString(txt, 1)
+			if r, err := regexp.Compile(slotRegex); err == nil {
+				match := r.FindAllString(question, 1)
 				if len(match) > 0 {
-					m.Slots[slot.Name] = match[0]
+					m.Slots[slotName] = match[0]
 				}
 			}
+		default:
```
Comment (ryancurrah) [inline fsm/fsm.go:235]: Added a default case here. Which makes `whole_text` the default matching type.

File: fsm/fsm_test.go [hunk 3]
```diff
@@ -1,128 +1,111 @@
...
+				State: 0,
+				Slots: make(map[string]string),
+			},
+			args: args{
+				cmd:       "turn_on",
+				txt:       "on",
+				fsmDomain: domainOnOff,
+			},
+			wantAnswers: []query.Answer{{
+				Text: "Turning on.",
```
Comment (ryancurrah) [inline fsm/fsm_test.go:91]: I cannot get this test case to return "Turning on.". I'm not sure what I'm doing wrong here. I'm digging into this currently. Trying to understand how it works.

File: cmd/cli/main.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package main
```
Comment (ryancurrah) [inline cmd/cli/main.go:1]: I separated the `chatto` server from the `cli`. So now we can run the cli locally and connect to the server remotely.

File: .goreleaser.yml [hunk 5]
```diff
@@ -5,7 +5,18 @@ before:
...
+  - id: chatto
+    binary: chatto
+    main: ./cmd/chatto/main.go
+    env:
+      - CGO_ENABLED=0
+    goos:
+      - linux
+      - windows
+      - darwin
+  - id: chatto-cli
```
Comment (ryancurrah) [inline .goreleaser.yml:17]: Publish a `chatto-cli` tool now.

File: internal/bot/server.go [hunk 6]
```diff
@@ -185,24 +192,24 @@ func (b *Bot) RegisterRoutes() {
 
 	r := mux.NewRouter()
 
-	// Channel endpoints
+	// Channel channels
 	if b.Channels.REST != nil {
-		r.HandleFunc("/endpoints/rest", b.restChannelHandler).Methods("POST")
+		r.HandleFunc("/channels/rest", b.restChannelHandler).Methods("POST")
```
Comment (ryancurrah) [inline internal/bot/server.go:197]: Renamed the endpoint URL from `endpoints` to `channels`.

**Comment by ryancurrah** [comment]
> Great I understand the concept of internal packages now, I had only seen it in a few projects before and I thought it was just a name, but it is very useful that Go itself uses the name to "hide" certain packages.
> 
> Of course, the `query` and `extension` packages (and `fsm` to some extent) should be public, in order to develop the extensions.
> 
> Also, some `bot` functions should be public, for example, `ServeBot`, that way you can build a "custom" bot, for example (this uses an old Chatto version):
> 
> ```go
> package main
> 
> import "github.com/jaimeteb/chatto/bot"
> 
> func main() {
>         path := "data/"
>         port := 4770
>         bot.ServeBot(&path, &port)
> }
> ```
> 
> Directory structure:
> 
> ```
> .
> ├── app.yaml
> ├── chatto_gae.go
> ├── data
> │   ├── clf.yml
> │   └── fsm.yml
> ├── go.mod
> └── go.sum
> ```

Ok I have implemented a `Client` and `Server` bot package.

**Comment by ryancurrah** [comment]
Ok @jaimeteb this PR is ready to go. I figured out why the test were failing. We have almost 100% coverage in the `fsm` package now.

File: fsm/fsm.go (hunk 2 above)
Comment (jaimeteb) [reply fsm/fsm.go:232]: Neat!

File: cmd/cli/main.go (hunk 4 above)
Comment (jaimeteb) [reply cmd/cli/main.go:1]: Very good idea

**Comment by jaimeteb** [comment]
> Ok @jaimeteb this PR is ready to go. I figured out why the test were failing. We have almost 100% coverage in the `fsm` package now.

That's amazing! I did a quick review, everything looks great, I just reverted `query.Answers` and a log in `fsm.TransitionStates`.

I'll do a few tests tomorrow and update the docs before merging.

**Comment by ryancurrah** [comment]
Thanks @jaimeteb !

**Comment by ryancurrah** [comment]
Oh I forgot I moved the log statement out of FSM and into the Bot. Generally it's best practice to not add log statements to libraries that other people import.

**Comment by jaimeteb** [comment]
> Oh I forgot I moved the log statement out of FSM and into the Bot. Generally it's best practice to not add log statements to libraries that other people import.

Oh you're right, I didn't notice that, I reverted that change.

**Pull request merged**